server/src/main/java/org/opensearch/cluster/service/MasterTaskThrottler.java

```diff
@@ -0,0 +1,153 @@
+        // Will be added as part of upcoming PRs,
+        // Right now just added "put-mapping" for ref only.
+        new HashSet<>(Arrays.asList("put-mapping"))
+    );
+
+    private final int DEFAULT_THRESHOLD_VALUE = -1; // Disabled throttling
+    private final MasterService masterService;
+    private final MasterTaskThrottlerListener masterTaskThrottlerListener;
+
+    private final ConcurrentMap<String, Long> tasksCount;
+    private final ConcurrentMap<String, Long> tasksThreshold;
+
+    public MasterTaskThrottler(
+        final ClusterSettings clusterSettings,
+        final MasterService masterService,
+        final MasterTaskThrottlerListener masterTaskThrottlerListener
+    ) {
+        clusterSettings.addSettingsUpdateConsumer(THRESHOLD_SETTINGS, this::updateSetting, this::validateSetting);
+        this.masterService = masterService;
+        this.masterTaskThrottlerListener = masterTaskThrottlerListener;
+        tasksCount = new ConcurrentHashMap<>(128); // setting initial capacity so each task will lend in different segment
+        tasksThreshold = new ConcurrentHashMap<>(128); // setting initial capacity so each task will lend in different segment
+    }
+
+    public void validateSetting(final Settings settings) {
+        /**
+         * TODO: Change the version number of check as per version in which this change will be merged.
+         */
+        if (masterService.state().nodes().getMinNodeVersion().compareTo(Version.V_3_0_0) < 0) {
+            throw new IllegalArgumentException("All the nodes in cluster should be on version later than or equal to 3.0.0");
+        }
+        Map<String, Settings> groups = settings.getAsGroups();
+        for (String key : groups.keySet()) {
+            if (!CONFIGURED_TASK_FOR_THROTTLING.contains(key)) {
+                throw new IllegalArgumentException("Master task throttling is not configured for given task type: " + key);
+            }
+            int threshold = groups.get(key).getAsInt("value", DEFAULT_THRESHOLD_VALUE);
+            if (threshold < DEFAULT_THRESHOLD_VALUE) {
+                throw new IllegalArgumentException("Provide positive integer for limit or -1 for disabling throttling");
+            }
+        }
+    }
+
+    public void updateSetting(final Settings settings) {
+        Map<String, Settings> groups = settings.getAsGroups();
+        for (String key : groups.keySet()) {
+            updateLimit(key, groups.get(key).getAsInt("value", DEFAULT_THRESHOLD_VALUE));
+        }
+    }
+
+    protected void updateLimit(final String taskKey, final int limit) {
+        assert limit >= DEFAULT_THRESHOLD_VALUE;
+        if (limit == DEFAULT_THRESHOLD_VALUE) {
+            tasksThreshold.remove(taskKey);
+        } else {
+            tasksThreshold.put(taskKey, (long)limit);
+        }
+    }
+
+    protected Long getThrottlingLimit(final String taskKey) {
+        return tasksThreshold.get(taskKey);
+    }
```

Comment:
Why protected? If this is for tests use package private

Reply:
Ack